Parser/lexer/lexer.c

```diff
@@ -0,0 +1,1419 @@
+#include "Python.h"
+#include "pycore_token.h"
+#include "pycore_unicodeobject.h"
+#include "errcode.h"
+
+#include "state.h"
+#include "../tokenizer/helpers.h"
+
+/* Alternate tab spacing */
+#define ALTTABSIZE 1
+
+#define is_potential_identifier_start(c) (\
+              (c >= 'a' && c <= 'z')\
+               || (c >= 'A' && c <= 'Z')\
+               || c == '_'\
+               || (c >= 128))
+
+#define is_potential_identifier_char(c) (\
+              (c >= 'a' && c <= 'z')\
+               || (c >= 'A' && c <= 'Z')\
+               || (c >= '0' && c <= '9')\
+               || c == '_'\
+               || (c >= 128))
+
+#ifdef Py_DEBUG
+static inline tokenizer_mode* TOK_GET_MODE(struct tok_state* tok) {
+    assert(tok->tok_mode_stack_index >= 0);
+    assert(tok->tok_mode_stack_index < MAXFSTRINGLEVEL);
+    return &(tok->tok_mode_stack[tok->tok_mode_stack_index]);
+}
+static inline tokenizer_mode* TOK_NEXT_MODE(struct tok_state* tok) {
+    assert(tok->tok_mode_stack_index >= 0);
+    assert(tok->tok_mode_stack_index + 1 < MAXFSTRINGLEVEL);
+    return &(tok->tok_mode_stack[++tok->tok_mode_stack_index]);
+}
+#else
+#define TOK_GET_MODE(tok) (&(tok->tok_mode_stack[tok->tok_mode_stack_index]))
+#define TOK_NEXT_MODE(tok) (&(tok->tok_mode_stack[++tok->tok_mode_stack_index]))
+#endif
+
+#define MAKE_TOKEN(token_type) _PyLexer_token_setup(tok, token, token_type, p_start, p_end)
+#define MAKE_TYPE_COMMENT_TOKEN(token_type, col_offset, end_col_offset) (\
+                _PyLexer_type_comment_token_setup(tok, token, token_type, col_offset, end_col_offset, p_start, p_end))
+
+/* Spaces in this constant are treated as "zero or more spaces or tabs" when
+   tokenizing. */
+static const char* type_comment_prefix = "# type: ";
+
+static inline int
+contains_null_bytes(const char* str, size_t size)
+{
+    return memchr(str, 0, size) != NULL;
+}
+
+/* Get next char, updating state; error code goes into tok->done */
+static int
+tok_nextc(struct tok_state *tok)
+{
+    int rc;
+    for (;;) {
+        if (tok->cur != tok->inp) {
+            tok->col_offset++;
+            return Py_CHARMASK(*tok->cur++); /* Fast path */
+        }
+        if (tok->done != E_OK) {
+            return EOF;
+        }
+        rc = tok->underflow(tok);
```

Comment:
@pablogsal This is the only change I did here. Instead of calling `underflow` according to stuff like `tok->str`, every tokenizer type passes in a callable that handles `underflow` in `_PyTokenizer_From*`.